File: apps/api/app/schemas/intelligence.py
"""
Schemas for the four endpoints the frontend's live provider expects but that
the original backend never shipped:

* GET /api/v1/persistent-sources   -> PersistentSourcesResponse
* GET /api/v1/watchtower           -> WatchtowerDigest
* GET /api/v1/analytics/density    -> DensityResponse
* GET /api/v1/events/{id}/report   -> EventReportResponse

These are *view* models — every value is derived from the same ThermalEvent /
IndustrialSite / Classification tables the rest of the API reads. That is
deliberate: it means the hotspot ranking, the heatmap and the per-event report
can never disagree with the event list, which is the exact inconsistency the
frontend was built to avoid.

Field names are camelCase-free (snake_case) on the wire; the frontend mapper
is responsible for the camelCase domain types in src/types/intelligence.ts.
"""

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class DensityCell(BaseModel):
    lat: float
    lon: float
    count: int
    mean_frp: float | None = None
    # One of classification_service.LABELS, or "other" when no classification
    # exists yet (the demo seed has none until /classify is called).
    dominant_class: str


class DensityResponse(BaseModel):
    bbox: list[float] = Field(..., description="[west, south, east, north]")
    cell_size_deg: float
    cells: list[DensityCell]
    generated_at: datetime


class PersistentSource(BaseModel):
    hotspot_id: int
    label: str
    kind: str = Field(..., description="industrial | wildfire | residue")
    state: str
    district: str
    lat: float
    lon: float
    active_days: int
    detection_count: int
    dominant_class: str
    max_frp: float | None = None
    avg_brightness: float | None = None
    priority_score: float
    facility_name: str | None = None
    facility_type: str | None = None
    distance_km: float | None = None
    first_date: str | None = None
    last_date: str | None = None


class PersistentSourcesResponse(BaseModel):
    window_days: int
    generated_at: datetime
    sources: list[PersistentSource]


class WatchtowerDigest(BaseModel):
    generated_at: datetime
    window_days: int
    new_events: int
    priority_events: int
    requires_review: int
    persistent_sources: int
    by_class: list[dict] = Field(..., description="[{category, count, percentage}]")
    top_regions: list[dict] = Field(..., description="[{state, count}]")


class EventReportResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    event_id: str
    generated_at: datetime
    classification: str
    classification_label: str
    confidence: float
    confidence_band: str
    priority_band: str
    priority_score: float
    location: dict
    persistence: dict
    thermal: dict
    nearest_facility: dict | None = None
    key_evidence: list[dict]
    caveats: list[str]
    provenance: dict

File: apps/web/src/types/intelligence.ts
/**
 * Derived intelligence types.
 *
 * These are *views* over `ThermalEvent[]`, never independent data. Everything
 * in here is computed by `src/data/derive.ts` from the one seeded dataset, so
 * the map, the analytics, the story and the report can never disagree with
 * each other — the exact failure the previous build had when it showed ten
 * dots under a headline claiming 12,543 events.
 */

import type {
  ThermalEvent,
  ThermalClass,
  Evidence,
  HistoricalObservation,
  PriorityBand,
  EventFilters,
} from './event';

/** A persistent thermal source, ranked. Powers the hotspot ranking feature. */
export interface PersistentSource {
  hotspotId: number;
  label: string;
  kind: 'industrial' | 'wildfire' | 'residue';
  state: string;
  district: string;
  lng: number;
  lat: number;
  activeDays: number;
  detectionCount: number;
  dominantClass: ThermalClass;
  maxFrp: number;
  avgBrightness: number;
  priorityScore: number;
  facilityName?: string;
  facilityType?: string;
  distanceKm?: number;
  firstDate: string;
  lastDate: string;
}

/** Everything the investigation panel needs about one event. */
export interface EventDetail {
  event: ThermalEvent;
  /** The persistent source this detection belongs to, if any. */
  source: PersistentSource | null;
  history: HistoricalObservation[];
  evidence: Evidence | null;
  priorityBand: PriorityBand;
  confidenceBand: 'high' | 'moderate' | 'uncertain';
  story: AnomalyStory;
  breadcrumb: string[];
}

/**
 * A concise structured narrative generated from structured fields. This is
 * template composition over measured values — not a chat model.
 */
export interface AnomalyStory {
  headline: string;
  sentences: string[];
  /** Ordered, most significant first. */
  drivers: Array<{ label: string; value: string; weight: number }>;
  caveats: string[];
}

/** Regional density cell for the thermal heatmap. */
export interface DensityCell {
  lng: number;
  lat: number;
  count: number;
  /** Mean fire radiative power in MW across the cell. */
  meanFrp: number;
  dominantClass: ThermalClass;
}

/** Watchtower digest — the monitoring view (new / priority / persistent). */
export interface WatchtowerDigest {
  generatedAt: string;
  windowDays: number;
  newEvents: ThermalEvent[];
  priorityEvents: ThermalEvent[];
  persistentSources: PersistentSource[];
  totals: {
    events: number;
    sources: number;
    persistentSources: number;
    industrial: number;
    requiresReview: number;
  };
}

/** Structured intelligence summary ready for export. */
export interface IntelligenceReport {
  eventId: string;
  generatedAt: string;
  classification: ThermalClass;
  classificationLabel: string;
  confidence: number;
  confidenceBand: string;
  priorityBand: PriorityBand;
  priorityScore: number;
  location: {
    lat: number;
    lng: number;
    state?: string;
    district?: string;
    breadcrumb: string[];
  };
  persistence: { activeDays: number; detectionCount: number; windowDays: number };
  thermal: { brightness: number; frp?: number; satellite: string; instrument: string; daynight: string };
  nearestFacility?: { name: string; type: string; distanceKm: number; bearingDeg?: number };
  keyEvidence: Array<{ factor: string; weight: number; detail: string; source: string }>;
  caveats: string[];
  provenance: {
    dataType: string;
    primarySource: string;
    satellites: string;
    modelVersion: string;
    industrialContext: string;
  };
}

/** Full analytics view derived from one dataset. */
export interface AnalyticsView {
  windowDays: number;
  period: { start: string; end: string };
  totals: {
    events: number;
    sources: number;
    persistentSources: number;
    industrialShare: number;
    requiresReview: number;
  };
  byClass: Array<{ class: ThermalClass; count: number; avgConfidence: number; share: number }>;
  byPriority: Array<{ band: PriorityBand; count: number }>;
  byState: Array<{ state: string; count: number }>;
  byDay: Array<{ date: string; count: number; industrial: number }>;
  bySatellite: Array<{ satellite: string; count: number }>;
  topSources: PersistentSource[];
}

export type { EventFilters };
